edx/analytics/tasks/insights/location_per_course.py

```diff
@@ -35,7 +35,7 @@ class LastIpAddressRecord(Record):
     """
     timestamp = StringField(description='Timestamp of last event by user in a course.')
     ip_address = StringField(description='IP address recorded on last event by user in a course.')
-    user_id = IntegerField(description='User ID recorded on last event by user in a course.')
+    username = StringField(description='Username recorded on last event by user in a course.')
```

Comment:
Can you provide some context on this change (and the rest of this commit)?

Why do we need this?
Does this mean edX doesn't have this change?
If not, should they?

Reply:
Ginkgo code used username from auth_user as the key. For Hawthorn they decided it was unnecessary to sqoop up the auth_user table and switched to using user_id, from the tracking log, as the key.

However, user_id is of no use to us because it breaks our unauthenticated user filtering code. The change reverts to the username method. 

Trying to add username as a field to filter on but retaining user_id as the key became very complicated since it isn't just a name change; the task of sqooping up auth_user is a pre-req. Additionally it would have meant re-running the history task since the change in data structure required a change in Hive store.

Reply:
Gotcha, thanks @dcadams.
I understand that we're using `username` as a
means of filtering out direct access users.
This leaves me with a few questions and concerns..

I'm concerned about forking the data model. I'm worried that this will
place us in an endless game of cat-and-mouse where each time we want to
upgrade Insights, we're forced to find and replace all uses of
`user_id`. I'm worried that essentially forking what is, in effect, a
primary key, will create lots of unnecessary work for us, to say nothing
of increased odds of bugs and regressions.


> they decided it was unnecessary to sqoop up the auth_user table

What did the team at edX say when you spoke to them about this?
It wouldn't surprise me to hear that following a refactor, they realized
that they no longer needed the `auth_user` dependency, as it wasn't be
referenced anywhere. I doubt it was either malice or carelessness.
This is, admittedly, a major reason why I requested we submit the
user-filtering code upstream; so that even if edX wasn't really using
it, they'd be aware of it and then less/un-likely to remove required
dependencies.
But as long as we keep forking code and keeping it to ourselves,
we should expect to keep having the rug pulled out from under us.

So that's a big-picture reason why this changeset concerns me;
- we made a small fork to add a feature
- by not submitting upstream, edx didn't know we needed certain code and
  it was removed.
- we now make a larger fork to further support the original feature...

Our experience has shown us this is likely to continue happening,
with us just digging deeper each time...


> Trying to add username as a field to filter on but retaining user_id as the key became very complicated

It's not clear to me why we couldn't have our cake and eat it too..
Why can't we leave `user_id` in place, using it everywhere edX already
expects to, while _also_ scooping up the `auth_user` table and then only
referencing `username` when we absolutely must, here: filtering direct
access users.

That said, if we _do_ need to fork this, scooping up `auth_user` and
filtering on usernames seems like a hack. It made sense when that was
the best piece of data available to us, but if we end up needing to fork
the pipeline to scoop up a table that edx doesn't already, shouldn't we
just get to the heart of the matter and scoop up `auth_userprofile`
instead and explicitly check the `nonregistered` flag?
Not only would it be more performant to do a bool/int comparison than a
wildcard string match, but it seems technically more correct to check
whether the user _truly_ is a direct access user, than to just check the
format of their username...

Regardless, since edx already does have `user_id`,
what's stopping us from just:
- using edx's `user_id` code everywhere
- also scooping up `auth_userprofile`
- and just filtering on `nonregistered == 1`


> it would have meant re-running the history task

How involved of a task is this?
I thought it was a fairly straight-forward procedure that we've
performed several times already.
Didn't we just do this back in December?
What more is involved beyond kicking off the job and waiting?

I'm curious to know how edX handled this;
I'm assuming they had to, right?
Since they made the same change, from `username` to `user_id`.

Given that their data set is significantly larger than ours,
if this transition worked for them, why doesn't it work for us?